### ISSUE TITLE
WinRM fails for specific tasks only

### ISSUE THREAD

**Issue by samalv** [ansible/ansible#70644]
##### SUMMARY

WinRM fails only for certain tasks while it works for others in same play using same modules

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
ansible_winrm_kerberos

##### ANSIBLE VERSION

```paste below

bash-4.4# ansible --version
ansible 2.9.9
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/var/lib/awx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.8 (default, Apr 16 2020, 01:36:27) [GCC 8.3.1 20191121 (Red Hat 8.3.1-5)]
```

##### CONFIGURATION

```paste below

bash-4.4# ansible-config dump --only-changed
bash-4.4#
```

##### OS / ENVIRONMENT

AWX official docker images

bash-4.4# cat /etc/redhat-release
CentOS Linux release 8.2.2004 (Core)

##### STEPS TO REPRODUCE

I wrote test play that run simple copy and Robo copy, it works for one folder and it failed for another, I tried also some other plays to run win_shell, so it works for one while it fails for another, I will paste the ones related to copy to avoid confusion with too many examples.

I tried also to upgrade dongrade pip module to various version with no luck, below are modules along with version during this run.

```
bash-4.4# pip list
Package                            Version
---------------------------------- ------------------
activedirectory                    0.14
ansible                            2.9.9
ansible-tower-cli                  3.3.9
asn1crypto                         0.24.0
awxkit                             12.0.0
Babel                              2.5.1
bcrypt                             3.1.6
certifi                            2020.6.20
cffi                               1.14.0
chardet                            3.0.4
click                              6.7
colorama                           0.4.3
cryptography                       2.9.2
dnspython                          1.16.0
enum34                             1.1.10
future                             0.18.2
gpg                                1.10.0
idna                               2.10
ijson                              2.6.1
importlib-resources                3.0.0
ipaddress                          1.0.23
Jinja2                             2.11.2
jmespath                           0.10.0
ldap3                              2.7
lxml                               4.5.2
MarkupSafe                         1.1.1
netaddr                            0.8.0
nsx-policy-python-sdk              2.5.1.0.5.16221899
nsx-python-sdk                     2.5.1.0.5.16221899
nsx-vmc-aws-integration-python-sdk 2.5.1.0.5.16221899
nsx-vmc-policy-python-sdk          2.5.1.0.5.16221899
ntlm-auth                          1.5.0
oauthlib                           3.1.0
omsdk                              1.2.445
orionsdk                           0.3.0
paramiko                           2.4.3
pip                                20.1.1
ply                                3.11
psycopg2                           2.7.5
pyasn1                             0.4.8
pycparser                          2.20
pycryptodomex                      3.9.8
pykerberos                         1.2.1
PyNaCl                             1.3.0
pyOpenSSL                          19.1.0
pysmi                              0.3.4
pysnmp                             4.4.12
pysnmp-mibs                        0.1.6
pysnow                             0.7.16
python-dateutil                    2.6.1
python-magic                       0.4.18
python-version                     0.0.2
pytz                               2019.3
pyvmomi                            7.0
pywinrm                            0.4.1
PyYAML                             5.3.1
requests                           2.24.0
requests-ntlm                      1.1.0
requests-oauthlib                  1.3.0
rpm                                4.14.2
selinux                            2.9
setuptools                         49.2.0
six                                1.15.0
suds-jurko                         0.6
supervisor                         4.2.0
urllib3                            1.25.9
vapi-client-bindings               3.3.0
vapi-common-client                 2.15.0
vapi-runtime                       2.15.0
virtualenv                         16.7.10
vmc-client-bindings                1.26.0
vmc-draas-client-bindings          1.9.0
vSphere-Automation-SDK             1.31.0
xmltodict                          0.12.0
zipp                               3.1.0
```

```yaml
- name: copy required pacakges
  hosts: "{{ hostvars['localhost']['sql_server_fqdn'] }}"
  gather_facts: yes
  tasks:
    - name: Robo-Copy MS SQL Server Binaries   <--- this works
      win_robocopy:
        src: "{{ pkg_server_root }}\\MSSQL2016Std_wSP2\MSSQL2016Ent_wSP2\x64\Setup\x64"
        dest: C:\SQLInstall\
        recurse: yes
      ignore_errors: yes

    - name: Copy MS SQL Server Binaries   <--- this works
      win_copy:
        src: "{{ pkg_server_root }}\\MSSQL2016Std_wSP2\autorun.inf"
        dest: C:\SQLInstall\
        remote_src: yes
        force: no
      ignore_errors: yes

    - name: Robo-Copy MS SQL Server Binaries   <--- this fails
      win_robocopy:
        src: '{{ pkg_server_root }}\\MSSQL2016Std_wSP2'
        dest: C:\SQLInstall\
        recurse: yes

    - name: Copy Binaries     <--- this fails
      win_copy:
        src: "{{ pkg_server_root }}\\MSSQL2016Std_wSP2"
        dest: C:\SQLInstall\
        remote_src: yes
        force: no
```

##### EXPECTED RESULTS

Files to be copied successfully 

##### ACTUAL RESULTS

I am getting the same error if I run certain win_copy or win_robocopy and also for some win_shell commands, while some other tasks usin same modules in same play just work fine. the ones that fail will always fail and the ones that work will always work

```paste below
{
    "msg": "Unexpected failure during module execution.",
    "exception": "Traceback (most recent call last):\n  
File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 239, in send_message\n    
root = ET.fromstring(ex.response_text)\n  File \"/usr/lib64/python3.6/xml/etree/ElementTree.py\", line 1315, in XML\n   
 return parser.close()\n  File \"<string>\", line None\nxml.etree.ElementTree.ParseError: no element found: line 1, column 0\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  
File \"/usr/local/lib/python3.6/site-packages/ansible/plugins/connection/winrm.py\", line 475, in _winrm_exec\n    resptuple = self.protocol.get_command_output(self.shell_id, command_id)\n  
File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 398, in get_command_output\n    self._raw_get_command_output(shell_id, command_id)\n
  File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 417, in _raw_get_command_output\n    res = self.send_message(xmltodict.unparse(req))\n  File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 242, in send_message\n    raise ex\n  File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 234, in send_message\n    resp = self.transport.send_message(message)\n  
File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/transport.py\", line 256, in send_message\n    response = self._send_message_request(prepared_request, message)\n  
File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/transport.py\", line 273, in _send_message_request\n    raise WinRMTransportError('http', ex.response.status_code, response_text)\nwinrm.exceptions.WinRMTransportError: Bad HTTP response returned from server. Code 400\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  
File \"/usr/local/lib/python3.6/site-packages/ansible/executor/task_executor.py\", line 146, in run\n    res = self._execute()\n  File \"/usr/local/lib/python3.6/site-packages/ansible/executor/task_executor.py\", line 645, in _execute\n    result = self._handler.run(task_vars=variables)\n  
File \"/usr/local/lib/python3.6/site-packages/ansible/plugins/action/normal.py\", line 46, in run\n    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))\n 
 File \"/usr/local/lib/python3.6/site-packages/ansible/plugins/action/__init__.py\", line 923, in _execute_module\n    res = self._low_level_execute_command(cmd, sudoable=sudoable, in_data=in_data)\n  File \"/usr/local/lib/python3.6/site-packages/ansible/plugins/action/__init__.py\", line 1075, in _low_level_execute_command\n    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)\n 
 File \"/usr/local/lib/python3.6/site-packages/ansible/plugins/connection/winrm.py\", line 548, in exec_command\n    result = self._winrm_exec(cmd_parts[0], cmd_parts[1:], from_exec=True, stdin_iterator=stdin_iterator)\n  
File \"/usr/local/lib/python3.6/site-packages/ansible/plugins/connection/winrm.py\", line 509, in _winrm_exec\n    self.protocol.cleanup_command(self.shell_id, command_id)\n 
 File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 372, in cleanup_command\n    res = self.send_message(xmltodict.unparse(req))\n  File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py\", line 234, in send_message\n    resp = self.transport.send_message(message)\n 
 File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/transport.py\", line 251, in send_message\n    prepared_request = self.encryption.prepare_encrypted_request(self.session, self.endpoint, message)\n  
File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/encryption.py\", line 79, in prepare_encrypted_request\n    encrypted_message = self._encrypt_message(message, host)\n  

File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/encryption.py\", line 108, in _encrypt_message\n    encrypted_stream = self._build_message(message, host)\n  
File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/encryption.py\", line 189, in _build_kerberos_message\n    sealed_message, signature = self.session.auth.wrap_winrm(host, message)\n  

File \"/var/lib/awx/venv/ansible/lib/python3.6/site-packages/requests_kerberos/kerberos_.py\", line 423, in wrap_winrm\n    return kerberos.authGSSWinRMEncryptMessage(self.context[host], message)\n

kerberos.GSSError: (('No context has been established', 524288), ('Attempt to use incomplete security context', 39756039))\n",
    "stdout": "",
    "_ansible_no_log": false
}
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jborean93** [comment]
This is very interesting, it hard to tell what is happening exactly but there are 3 exceptions so stepping through each one. The first exception is

```python
Traceback (most recent call last):
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 239, in send_message
    
root = ET.fromstring(ex.response_text)
  File "/usr/lib64/python3.6/xml/etree/ElementTree.py", line 1315, in XML
   
 return parser.close()
  File "<string>", line None
xml.etree.ElementTree.ParseError: no element found: line 1, column 0
```

We can see in the code it tries to [parse the response content](https://github.com/diyan/pywinrm/blob/667703f3ddebc67cab72d86620315c739bfd7262/winrm/protocol.py#L248) and if an exception occurs here (which it does) it raises the original transport exception which is

```python
Traceback (most recent call last):
  
File "/usr/local/lib/python3.6/site-packages/ansible/plugins/connection/winrm.py", line 475, in _winrm_exec
    resptuple = self.protocol.get_command_output(self.shell_id, command_id)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 398, in get_command_output
    self._raw_get_command_output(shell_id, command_id)

  File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 417, in _raw_get_command_output
    res = self.send_message(xmltodict.unparse(req))
  File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 242, in send_message
    raise ex
  File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 234, in send_message
    resp = self.transport.send_message(message)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/transport.py", line 256, in send_message
    response = self._send_message_request(prepared_request, message)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/transport.py", line 273, in _send_message_request
    raise WinRMTransportError('http', ex.response.status_code, response_text)
winrm.exceptions.WinRMTransportError: Bad HTTP response returned from server. Code 400
```

This shows that a WinRM shell had been created but when Ansible went to start a process it failed to send the data needed to kick it off and WinRM returned the 400 error. From there Ansible is somewhat kicking off a new module which results in the traceback:

```python
Traceback (most recent call last):
  
File "/usr/local/lib/python3.6/site-packages/ansible/executor/task_executor.py", line 146, in run
    res = self._execute()
  File "/usr/local/lib/python3.6/site-packages/ansible/executor/task_executor.py", line 645, in _execute
    result = self._handler.run(task_vars=variables)
  
File "/usr/local/lib/python3.6/site-packages/ansible/plugins/action/normal.py", line 46, in run
    result = merge_hash(result, self._execute_module(task_vars=task_vars, wrap_async=wrap_async))
 
 File "/usr/local/lib/python3.6/site-packages/ansible/plugins/action/__init__.py", line 923, in _execute_module
    res = self._low_level_execute_command(cmd, sudoable=sudoable, in_data=in_data)
  File "/usr/local/lib/python3.6/site-packages/ansible/plugins/action/__init__.py", line 1075, in _low_level_execute_command
    rc, stdout, stderr = self._connection.exec_command(cmd, in_data=in_data, sudoable=sudoable)
 
 File "/usr/local/lib/python3.6/site-packages/ansible/plugins/connection/winrm.py", line 548, in exec_command
    result = self._winrm_exec(cmd_parts[0], cmd_parts[1:], from_exec=True, stdin_iterator=stdin_iterator)
  
File "/usr/local/lib/python3.6/site-packages/ansible/plugins/connection/winrm.py", line 509, in _winrm_exec
    self.protocol.cleanup_command(self.shell_id, command_id)
 
 File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 372, in cleanup_command
    res = self.send_message(xmltodict.unparse(req))
  File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/protocol.py", line 234, in send_message
    resp = self.transport.send_message(message)
 
 File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/transport.py", line 251, in send_message
    prepared_request = self.encryption.prepare_encrypted_request(self.session, self.endpoint, message)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/encryption.py", line 79, in prepare_encrypted_request
    encrypted_message = self._encrypt_message(message, host)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/encryption.py", line 108, in _encrypt_message
    encrypted_stream = self._build_message(message, host)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/winrm/encryption.py", line 189, in _build_kerberos_message
    sealed_message, signature = self.session.auth.wrap_winrm(host, message)
  
File "/var/lib/awx/venv/ansible/lib/python3.6/site-packages/requests_kerberos/kerberos_.py", line 423, in wrap_winrm
    return kerberos.authGSSWinRMEncryptMessage(self.context[host], message)

kerberos.GSSError: (('No context has been established', 524288), ('Attempt to use incomplete security context', 39756039))
```

This error is likely due to the 2nd traceback breaking the security context and when it goes to use it it will fail. Why this has happened I'm not sure, each new task should be using a new context. The only exception to this would be using a loop but I don't see any of those in your example.

You said the same module works fine in some scenarios but in others it fails. Does the same module with options also fail if you run it by yourself, either as a single task in a playbook or through just an ansible call? It would also be good to do testing on a simple module like win_shell as win_copy is an action plugin that can run multiple modules making it hard to see the error.

Also just an FYI, when you use double quote for a value you need to be careful of backslashes. When used in a double quote they become an escape character so if you do `"C:\temp\new_file.txt"` then that becomes `C:<tab>emp<new line>ew_file.txt` which is not what you want. I usually recommend you use single quotes when you need to quote a yaml value as it won't treat `\` as an escape char but if you want to continue using double quotes just double up on every backslash like `"C:\\temp\\new_file.txt"`.

**Comment by samalv** [comment]
Thanks @jborean93 

I tried this morning to run the same play in a command line Ansible (outside AWX) and it worked :-) not sure how AWX is different comparing to command line Ansible (both are running same version of Ansible)

**Comment by jborean93** [comment]
In this case AWX is just calling Ansible which is where all the tracebacks are occurring so it's not something it's doing directly. Were you able to run each task in AWX as separate tasks rather than altogether? You can try running Ansible through the AWX venv to see if that's causing some issues. I don't know too much about AWX but you could try disabling job isolation in case that's causing a problem but I doubt it.

Unfortunately this is going to be difficult to track down, to really continue forward we need a way to replicate the problem somehow and to try and reduce the producer to as little moving parts as possible. The first step would be to try and figure out if there is a single task that we can run that always causes the problem. From there we can try different auth methods like `ntlm` instead of Kerberos. We could also try the [psrp](https://docs.ansible.com/ansible/latest/plugins/connection/psrp.html) connection plugin to see if that's also affected.

**Comment by samalv** [comment]
Hello again,

I just figured it out, a colleague of mine has configured proxy in AWX, there was a mistake within (no_proxy), I just removed that and then it works. In addition it seems we have also serious problem with proxy and that's why it is broken. I can not explain while some of the tasks are failing while some are working towards same destination using same modules. But it is clear the issue is proxy related.

Thanks indeed guys for your help ... it really helped me a lot tracking this down.
